### DISCUSSION TITLE
Is OpenAPI 3.1.0 supported  ?

### DISCUSSION THREAD

**Discussion by gautamnabin5** [RicoSuter/NSwag#4599] [Q&A]
Was wondering if .NET core code generation using open api spec file version 3.1.0 is supported ?

**Comment by stevenvolckaert** [comment]
I'm having the same question.

@RicoSuter is [OpenAPI 3.1.0](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md) supported by the latest [NSwag.MSBuild](https://www.nuget.org/packages/NSwag.MSBuild/) version, v13.16.0?

Thank you!

**Comment by SimonCull** [comment]
Also interested in this.

**Comment by jhancock4d** [comment]
For me, I'm specifically interested in OpenAPI 3.1 for the following cases:

1. Defininig enums using OneOf and having maps of values to descriptive text.
2. Response generics so that we can return things like an IAsyncEnumerable of items of whatever type we want, and say a totalCount and a few other things and not have it generate a bazillion different types for no reason.

**Comment by glyad** [comment]
Waiting for an answer...

**Comment by amrud** [comment]
interested in this too...

**Comment by trenslow** [comment]
did anyone get an answer for this?

**Comment by dlgombert** [comment]
Hi.  Please tell.

**Comment by andrewmshea** [comment]
Let me know :(

**Comment by amilos** [comment]
Answer would be appreciated :)

**Comment by meiedomi** [comment]
We are currently evaluating NSwag for our new project using OpenAPI 3.1 too, but I couldn't find any official information regarding the support for 3.1. Any updates?

**Comment by lahma** [comment]
If there's someone on this thread being both technically apt and having understanding of the 3.1 spec it might be pretty useful to test it against v14 preview and report what is missing instead of asking if someone could test it on your behalf. If it helps I can always start to ping everyone asking about "any updates with testing it?" 🙄

**Comment by JohnGalt1717** [comment]
Doesn't appear to move the ball forward at all on any of these things.

Enums need oneOf support
Generics need to be supported (which right now just throw a horrible error message that tells you nothing)

If someone needs it, here's a SchemaFilter for the enums support:

```
public sealed class FixNSwagEnumsProcessor : ISchemaProcessor {
	public void Process(SchemaProcessorContext context) {
		if (!context.ContextualType.Type.IsEnum)
			return;
		
		foreach(var item in context.Schema.EnumerationNames.Index()) {
			var enumValue = context.Schema.Enumeration.ElementAt(item.index);
			var enumSchema = new JsonSchema
			{
				ExtensionData = new Dictionary<string, object?>
				{
					{ "const", enumValue },
					{ "title", item.item }
				}
			};

			context.Schema.OneOf.Add(enumSchema);			
		}
	}
}
```

But generics is a hard break. And it doesn't even do what Swashbuckle does and generate a whole pile of unique ref types for every generic it finds.

**Comment by subhajit-aec** [comment]
I am using NSwag.Core (14.0.8) and NSwag.Core.Yaml (14.0.8) but getting a **Newtonsoft.Json.JsonSerializationException.**

**Yml file 👇**
[test.yml.txt](https://github.com/user-attachments/files/16049787/test.yml.txt)

**This is my code 👇**
![image](https://github.com/RicoSuter/NSwag/assets/141990144/dc1833b0-33bc-4f86-a1ed-c6462aacf5eb)

Does the latest version (14.0.8) support OpenAPI version 3.1.0?

**Comment by AnnaVernerovaHID** [comment]
We're using NSwag for generating swagger from C# code. **When generating schemas for objects based on C# classes with inheritance, the resulting swagger contains schemas specifying contradictory conditions**, like this:

```
"SchemaOfChildClass": {
  "allOf": [
    {
      "$ref": "#/components/schemas/SchemaOfParentClass"
    },
    {
      "type": "object",
      "additionalProperties": false, // <--- THIS IS THE CULPRIT
      "properties": {
        "foo": { "type": "string" }
      }
    }
  ]
},
"SchemaOfParentClass": {
  "type": "object",
  "additionalProperties": false,
  "properties": {
    "bar": { "type": "string" }
  }
}
```

Notice that property `bar` is required by the `SchemaOfParentClass` branch of `allOf` but, due to `"additionalProperties": false`, it is simultaneously prohibited by the spelled-out branch that lists properties added in the `ChildClass`. This is a known problem with OpenAPI specification version 3.0. In version 3.1, this would be properly solved by using `"unevaluagedProperties": false` on `SchemaOfParentClass` itself.
